File: odbc/check.py
"""Driver interoperability check for ExecDB's pgwire (spec §8, phase 4
follow-up): psqlODBC (the official PostgreSQL ODBC driver, via pyodbc),
against a running ExecDB instance already seeded with table t(a INTEGER)
(seeded by run-all.sh below).

Unlike the other 5 verified drivers, psqlODBC cannot connect with zero
ExecDB-side support: its connection bootstrap queries the real Postgres
system catalog (pg_type) to check for large-object support, and
SQLTables()/SQLColumns() (used here, and by real schema-browsing ODBC
consumers like Excel/Power BI/Access) query pg_class/pg_namespace/
pg_attribute/pg_attrdef plus the pg_get_expr()/current_schema() built-in
functions. ExecDB answers all of these via a small pg_catalog-compatible
set of temp views/functions set up per pgwire connection
(cmd/execdb/pgcatalog.go) -- see that file's doc comments for exactly
what is (and is not) emulated.
"""
import sys

import pyodbc


def main():
    if len(sys.argv) != 2:
        print("usage: check.py <connection-string>", file=sys.stderr)
        sys.exit(2)

    conn = pyodbc.connect(sys.argv[1], autocommit=True)
    try:
        check(conn)
    finally:
        conn.close()
    print("OK")


def check(conn):
    cur = conn.cursor()

    cur.execute("SELECT ?", (1,))
    (one,) = cur.fetchone()
    assert str(one) == "1", f"SELECT ? (1) returned {one!r}"

    cur.execute("SELECT ?", (3.5,))
    (f,) = cur.fetchone()
    assert float(f) == 3.5, f"SELECT ? (3.5) returned {f!r}"

    cur.execute("SELECT ?", ("hello",))
    (s,) = cur.fetchone()
    assert s == "hello", f"SELECT ? ('hello') returned {s!r}"

    cur.execute("SELECT NULL")
    (null,) = cur.fetchone()
    assert null is None, f"SELECT NULL returned {null!r}"

    # spec §2: DDL must be rejected via the external I/F, and pyodbc must
    # surface it as an Error with SQLSTATE 42501 (spec §8).
    try:
        cur.execute("CREATE TABLE odbc_should_not_exist(a INTEGER)")
        raise AssertionError("expected CREATE TABLE to be rejected via the external I/F")
    except pyodbc.Error as e:
        assert e.args[0] == "42501", f"expected SQLSTATE 42501, got {e.args[0]!r} ({e})"

    # A basic write/read round trip against the table run-all.sh seeded.
    cur.execute("INSERT INTO t VALUES (777006)")
    cur.execute("SELECT count(*) FROM t WHERE a = 777006")
    (n,) = cur.fetchone()
    assert n == 1, f"expected count=1 after INSERT, got {n}"

    # cmd/execdb/pgcatalog.go's actual purpose: a driver's own schema
    # browser (SQLTables/SQLColumns) sees ExecDB's real, current schema.
    tables = [row.table_name for row in cur.tables() if row.table_type == "TABLE"]
    assert "t" in tables, f"cursor.tables() did not list 't': {tables!r}"

    columns = {row.column_name: row.type_name for row in cur.columns(table="t")}
    assert "a" in columns, f"cursor.columns(table='t') did not list 'a': {columns!r}"


if __name__ == "__main__":
    main()

File: node/check.js
// Driver interoperability check for ExecDB's pgwire (spec §8, phase 4
// Step 7): node-postgres (pg), used with its own default connection
// settings (no ExecDB-specific workaround flags), against a running
// ExecDB instance already seeded with table t(a INTEGER) (seeded by
// run-all.sh below).
//
// The parameterized queries below ($1) make pg send Parse/Bind/Describe/
// Execute/Sync (Extended Query, phase 4 Step 5) rather than a plain Query
// message, exercising that path from a second driver besides pgx
// (tests/pgclient already covers pgx's default Extended Query mode).
'use strict';

const { Client } = require('pg');

async function main() {
  const dsn = process.argv[2];
  if (!dsn) {
    console.error('usage: node check.js <connectionString>');
    process.exit(2);
  }

  const client = new Client({ connectionString: dsn });
  await client.connect();
  try {
    await check(client);
  } finally {
    await client.end();
  }
  console.log('OK');
}

function firstValue(result) {
  return Object.values(result.rows[0])[0];
}

async function check(client) {
  let res = await client.query('SELECT $1', [1]);
  const one = firstValue(res);
  // int8 (OID 20, ExecDB's mapping for an INTEGER-affinity column,
  // cmd/execdb/pgtype.go) is returned by pg as a string by default --
  // JS's Number can't safely represent the full int8 range -- so compare
  // as a string rather than assuming a JS number.
  if (String(one) !== '1') {
    throw new Error(`SELECT $1 (1) returned ${JSON.stringify(one)}`);
  }

  res = await client.query('SELECT $1', [3.5]);
  const f = firstValue(res);
  if (Number(f) !== 3.5) {
    throw new Error(`SELECT $1 (3.5) returned ${JSON.stringify(f)}`);
  }

  res = await client.query('SELECT $1', ['hello']);
  const s = firstValue(res);
  if (s !== 'hello') {
    throw new Error(`SELECT $1 ('hello') returned ${JSON.stringify(s)}`);
  }

  res = await client.query("SELECT x'00ff'");
  const blob = firstValue(res);
  if (!Buffer.isBuffer(blob) || !blob.equals(Buffer.from([0x00, 0xff]))) {
    throw new Error(`SELECT x'00ff' returned ${JSON.stringify(blob)}`);
  }

  res = await client.query('SELECT NULL');
  const nullVal = firstValue(res);
  if (nullVal !== null) {
    throw new Error(`SELECT NULL returned ${JSON.stringify(nullVal)}`);
  }

  // spec §2: DDL must be rejected via the external I/F, and pg must
  // surface it as an error carrying SQLSTATE 42501 (spec §8).
  let rejected = false;
  try {
    await client.query('CREATE TABLE node_should_not_exist(a INTEGER)');
  } catch (e) {
    rejected = true;
    if (e.code !== '42501') {
      throw new Error(`expected SQLSTATE 42501 for the rejected CREATE TABLE, got ${e.code} (${e.message})`);
    }
  }
  if (!rejected) {
    throw new Error('expected CREATE TABLE to be rejected via the external I/F');
  }

  // A basic write/read round trip against the table run-all.sh seeded.
  await client.query('INSERT INTO t VALUES ($1)', [777002]);
  res = await client.query('SELECT count(*) FROM t WHERE a = $1', [777002]);
  const n = Number(firstValue(res));
  if (n !== 1) {
    throw new Error(`expected count=1 after INSERT, got ${n}`);
  }
}

main().catch((err) => {
  console.error('Error:', err.message || err);
  process.exit(1);
});
